rcl/include/rcl/publisher.h

```diff
@@ -0,0 +1,277 @@
+ */
+RCL_PUBLIC
+rcl_ret_t
+rcl_publish(const rcl_publisher_t * publisher, const void * ros_message);
+
+/// Get the topic name for the publisher.
+/* This function returns the publisher's internal topic name string.
+ * This function can fail, and therefore return NULL, if the:
+ *   - publisher is NULL
+ *   - publisher is invalid (never called init, called fini, or invalid node)
+ *
+ * The returned string is only valid as long as the rcl_publisher_t is valid.
+ * The value of the string may change if the topic name changes, and therefore
+ * copying the string is recommended if this is a concern.
+ *
+ * This function is not thread-safe, and copying the result is not thread-safe.
+ * \TODO(wjwwood): should we provide a thread-safe copy_topic_name?
+ *
+ * \param[in] publisher pointer to the publisher
+ * \return name string if successful, otherwise NULL
+ */
+RCL_PUBLIC
+const char *
+rcl_publisher_get_topic_name(const rcl_publisher_t * publisher);
+
+/// Return the rcl publisher options.
+/* This function returns the publisher's internal options struct.
+ * This function can fail, and therefore return NULL, if the:
+ *   - publisher is NULL
+ *   - publisher is invalid (never called init, called fini, or invalid node)
+ *
+ * The returned struct is only valid as long as the rcl_publisher_t is valid.
+ * The values in the struct may change if the options of the publisher change,
+ * and therefore copying the struct is recommended if this is a concern.
+ *
+ * This function is not thread-safe, and copying the result is not thread-safe.
+ * \TODO(wjwwood): should we provide a thread-safe copy_publisher_options?
+ *
+ * \param[in] publisher pointer to the publisher
+ * \return options struct if successful, otherwise NULL
+ */
+RCL_PUBLIC
+const rcl_publisher_options_t *
+rcl_publisher_get_options(const rcl_publisher_t * publisher);
+
+/// Return the rmw publisher handle.
+/* The handle returned is a pointer to the internally held rmw handle.
+ * This function can fail, and therefore return NULL, if the:
+ *   - publisher is NULL
+ *   - publisher is invalid (never called init, called fini, or invalid node)
+ *
+ * The returned handle is only valid as long as the given publisher is valid.
+ *
+ * \TODO(wjwwood) will the publisher ever need to remove and recreate the rmw
+ *                handle under the hood? Perhaps not now, but if we add the
+ *                ability to reconfigure, alias, or remap topics, and don't
+ *                implement that into rmw, then we should consider the
+ *                implications here. This argument applies to other parts of
+ *                the interface, but I'll just mention it here.
+ *
+ * \TODO(wjwwood) should the return value of this be const?
+ *
+ * \param[in] publisher pointer to the rcl publisher
+ * \return rmw publisher handle if successful, otherwise NULL
+ */
+RCL_PUBLIC
+rmw_publisher_t *
+rcl_publisher_get_rmw_publisher_handle(const rcl_publisher_t * publisher);
```

Comment:
Should we shorten the function name to not repeat `publisher`: `rcl_publisher_get_rmw_handle`?